python/ray/data/tests/test_dataset_pipeline.py

```diff
@@ -445,6 +446,69 @@ def test_count_sum_on_infinite_pipeline(ray_start_regular_shared):
     assert 9 == pipe.sum()
 
 
+def test_iter_batches_no_spilling_for_repeat(shutdown_only):
+    # The object store is about 300MB.
+    ctx = ray.init(num_cpus=1, object_store_memory=300e6)
+    # The size of dataset is 500*(80*80*4)*8B, about 100MB.
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100)
+
+    # Repeating the 100MB dataset 10 times should not explode object store.
+    pipe = ds.repeat(10)
+    for batch in pipe.iter_batches():
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.repeat(10)
+    for batch in pipe.iter_batches(prefetch_blocks=5):
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.map_batches(lambda x: x).repeat(10)
+    for batch in pipe.iter_batches():
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.map_batches(lambda x: x).repeat(10)
+    for batch in pipe.iter_batches(prefetch_blocks=5):
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+
+def test_iter_batches_no_spilling_for_window(shutdown_only):
+    # The object store is about 300MB.
+    ctx = ray.init(num_cpus=1, object_store_memory=300e6)
+    # The size of dataset is 500*(80*80*4)*8B, about 100MB.
+    ds = ray.data.range_tensor(500, shape=(80, 80, 4), parallelism=100)
+
+    pipe = ds.window(blocks_per_window=20).repeat(10)
+    for batch in pipe.iter_batches():
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.window(blocks_per_window=20).repeat(10)
+    for batch in pipe.iter_batches(prefetch_blocks=5):
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.map_batches(lambda x: x).window(blocks_per_window=20).repeat(10)
+    for batch in pipe.iter_batches():
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+    pipe = ds.map_batches(lambda x: x).window(blocks_per_window=20).repeat(10)
+    for batch in pipe.iter_batches(prefetch_blocks=5):
+        pass
+    meminfo = memory_summary(ctx.address_info["address"], stats_only=True)
+    assert "Spilled" not in meminfo, meminfo
+
+
```

Comment:
Also add tests for (1) map_batches after repeat, and (2) random_shuffle_each_window after repeat?

Reply:
Yep, added.